HeterogeneousCore/CUDAServices/src/CUDAService.cc

```diff
@@ -39,6 +52,60 @@ CUDAService::CUDAService(edm::ParameterSet const& config, edm::ActivityRegistry&
   enabled_ = true;
+
+  // see the documentation of cudaDeviceSetLimit for more information
+  auto limits = config.getUntrackedParameter<edm::ParameterSet>("limits");
+  size_t request, value;
+  for (int i = 0; i < numberOfDevices_; ++i) {
+    cudaCheck(cudaSetDevice(i));
+    // cudaLimitPrintfFifoSize controls the size in bytes of the shared FIFO used by the
+    // printf() device system call.
+    request = limits.getUntrackedParameter<unsigned int>("printfFifoSize");
+    if (request > 0) {
+      cudaCheckDeviceSetLimit(i, cudaLimitPrintfFifoSize, "cudaLimitPrintfFifoSize", request);
+    }
+    // cudaLimitStackSize controls the stack size in bytes of each GPU thread.
+    request = limits.getUntrackedParameter<unsigned int>("stackSize");
+    if (request > 0) {
+      cudaCheckDeviceSetLimit(i, cudaLimitStackSize, "cudaLimitStackSize", request);
+    }
+    // cudaLimitMallocHeapSize controls the size in bytes of the heap used by the malloc()
+    // and free() device system calls.
+    request = limits.getUntrackedParameter<unsigned int>("mallocHeapSize");
+    if (request > 0) {
+      cudaCheckDeviceSetLimit(i, cudaLimitMallocHeapSize, "cudaLimitMallocHeapSize", request);
+    }
+    // cudaLimitDevRuntimeSyncDepth controls the maximum nesting depth of a grid at which
+    // a thread can safely call cudaDeviceSynchronize().
+    request = limits.getUntrackedParameter<unsigned int>("devRuntimeSyncDepth");
+    if (request > 0) {
+      cudaCheckDeviceSetLimit(i, cudaLimitDevRuntimeSyncDepth, "cudaLimitDevRuntimeSyncDepth", request);
+    }
+    // cudaLimitDevRuntimePendingLaunchCount controls the maximum number of outstanding
+    // device runtime launches that can be made from the current device.
+    request = limits.getUntrackedParameter<unsigned int>("devRuntimePendingLaunchCount");
+    if (request > 0) {
+      cudaCheckDeviceSetLimit(i, cudaLimitDevRuntimePendingLaunchCount, "cudaLimitDevRuntimePendingLaunchCount", request);
+    }
+  }
+
+  for (int i = 0; i < numberOfDevices_; ++i) {
+    size_t value;
+    cudaCheck(cudaSetDevice(i));
+    std::cerr << "CUDA device " << i << ': \n';
+    cudaCheck(cudaDeviceGetLimit(&value, cudaLimitPrintfFifoSize));
+    std::cerr << "  cudaLimitPrintfFifoSize               " << value << '\n';
+    cudaCheck(cudaDeviceGetLimit(&value, cudaLimitStackSize));
+    std::cerr << "  cudaLimitStackSize                    " << value << '\n';
+    cudaCheck(cudaDeviceGetLimit(&value, cudaLimitStackSize));
+    std::cerr << "  cudaLimitMallocHeapSize               " << value << '\n';
+    cudaCheck(cudaDeviceGetLimit(&value, cudaLimitMallocHeapSize));
+    std::cerr << "  cudaLimitDevRuntimeSyncDepth          " << value << '\n';
+    cudaCheck(cudaDeviceGetLimit(&value, cudaLimitDevRuntimeSyncDepth));
+    std::cerr << "  cudaLimitDevRuntimePendingLaunchCount " << value << '\n';
+    cudaCheck(cudaDeviceGetLimit(&value, cudaLimitDevRuntimePendingLaunchCount));
+    std::cerr << std::endl;
+  }
```

Comment:
I this loop still needed?

Reply:
definitely not... looks like I pushed (to) the wrong branch